### PR TITLE
Replace JNI & C with jnr-ffi

### PR DESCRIPTION
**Pull request by hito4t**
Performance test is done.
Almost same as it used to.

### CODE REVIEW

File: embulk-output-oracle/src/main/java/org/embulk/output/oracle/oci/RowBuffer.java [hunk 1]
```diff
@@ -5,33 +5,45 @@
...
-            rowSize += column.columnSize;
+        for (int i = 0; i < table.getColumnCount(); i++) {
+            rowSize += table.getColumn(i).getDataSize();
         }
 
-        buffer = new byte[rowSize * rowCount];
+        // should be direct because used by native library
+        buffer = ByteBuffer.allocateDirect(rowSize * rowCount);
+        // position is not updated
+        defaultBuffer = buffer.duplicate();
```
Comment (frsyuki) [inline embulk-output-oracle/src/main/java/org/embulk/output/oracle/oci/RowBuffer.java]: :+1:

File: embulk-output-oracle/src/main/java/org/embulk/output/oracle/DirectBatchInsert.java [hunk 2]
```diff
@@ -97,9 +99,9 @@ public void prepare(String loadTable, JdbcSchema insertSchema) throws SQLExcepti
                 case Types.LONGNVARCHAR:
                 case Types.NCLOB:
                     columns.add(new ColumnDefinition(insertColumn.getName(),
-                            ColumnDefinition.SQLT_CHR,
+                    OCI.SQLT_CHR,
```
Comment (frsyuki) [inline embulk-output-oracle/src/main/java/org/embulk/output/oracle/DirectBatchInsert.java]: indent seems wrong. you need 8 more spaces :-)

**Comment by frsyuki** [comment]
I read OCIManager and it seems tricky because DirectBatchInsert owns it at a static field. Does this new code still need it?
If my understanding is correct, you can move `public OCIWrapper OCIManager.open(key, dbName, ...)` to `public static OCIWrapper OCIWrapper.open(dbName, ...)`. Then, DirectBatchInsert can own OCIWrapper instance instead of `ociKey).

**Comment by frsyuki** [comment]
oh, perhaps is OCIManager used to reduce number of OCIWrapper instances by sharing it across all threads? Then, using static field sounds OK (although using WeakHashMap is safer in this case but implementation will be complicated).

Even then, OCIWrapper.loadLibrary shouldn't be called multiple times, correct? I think you can use following idiom:

```
public class OCIWrapper
{
    private static OCI oci;
    private static Throwable libraryLoadError;

    static {
        try {
            oci = loadLibrary();
            libraryLoadError = null;
        }
        catch (Error | Exception ex) {
            // exception will be thrown later at getLibrary
            libraryLoadError = ex;
        }
    }

    private static OCI loadLibrary()
    {
        // ... (current OCIWrapper.loadLibrary code here) ...
    }

    public OCI getLibrary()
    {
        if (libraryLoadError != null) {
            throw Throwables.propagate(libraryLoadError);
        }
        return oci;
    }

    public OCIWrapper()
    {
        // ...
        oci = getLibrary();
    }
}
```

**Comment by hito4t** [comment]
Thank you for comments!

> indent seems wrong. you need 8 more spaces :-)

I've corrected.

> oh, perhaps is OCIManager used to reduce number of OCIWrapper instances (=number of database connections) by sharing it across all threads?

I would rather intend to load records in the same transaction.

Anyway, I've changed `oci` field into static not to call `loadLibrary` multiple times.

**Pull request merged**